### PR TITLE
[CB] Align the speculative decoding sample with other ones

### PR DESCRIPTION
**Pull request by iefode**
Spin off of https://github.com/openvinotoolkit/openvino.genai/pull/907. Merge after head PR

### CODE REVIEW

File: src/cpp/src/timer.hpp [hunk 1]
```diff
@@ -31,6 +31,6 @@ class ManualTimer {
     }
 
     ~ManualTimer() {
-        std::cout << m_title << ": " << m_total / 1000. << " secs" << std::endl;
+        // std::cout << m_title << ": " << m_total / 1000. << " secs" << std::endl;
```
Comment (andrei-kochin) [inline src/cpp/src/timer.hpp:34]: Remove?

Comment (iefode) [reply src/cpp/src/timer.hpp:34]: Decided to comment for debug purposes, best decision is move such prints under `Debug` flag

File: samples/CMakeLists.txt [hunk 2]
```diff
@@ -38,6 +39,7 @@ install(DIRECTORY
             python/greedy_causal_lm
             python/multinomial_causal_lm
             python/whisper_speech_recognition
+            cpp/speculative_decoding_lm
```
Comment (ilya-lavrenov) [inline samples/CMakeLists.txt:42]: this is only for python samples. You have installed `cpp/speculative_decoding_lm` above.

File: samples/cpp/speculative_decoding_lm/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # Copyright (C) 2024 Intel Corporation
+# Copyright (C) 2024 Intel Corporation
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/CMakeLists.txt:2]: duplicate

File: samples/cpp/speculative_decoding_lm/CMakeLists.txt [hunk 4]
```diff
@@ -21,14 +22,25 @@ if(NOT TARGET nlohmann_json)
     FetchContent_MakeAvailable(nlohmann_json)
 endif()
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/CMakeLists.txt:23]: I think we need to drop those dependencies are they are not used in current version of sample

File: samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp [hunk 5]
```diff
@@ -2,108 +2,22 @@
...
 
-    const size_t num_prompts = result["num_prompts"].as<size_t>();
-    const bool dynamic_split_fuse = result["dynamic_split_fuse"].as<bool>();
-    const std::string model_path = result["model"].as<std::string>();
-    const std::string draft_model_path = result["draft_model"].as<std::string>();
-    const std::string device = result["device"].as<std::string>();
-    const bool use_prefix = result["use_prefix"].as<bool>();
+    ov::genai::GenerationConfig config;
+    config.max_new_tokens = 100;
+    config.num_assistant_tokens = 5;
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp:15]: you can add a comment above that users are encouraged to use other values depending of main / draft model or even new mode, which is based on probabilities

Or add such information about 2 modes in readme.md

File: samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp [hunk 6]
```diff
@@ -116,34 +30,25 @@ int main(int argc, char* argv[]) try {
...
+    scheduler_config.cache_size = 5;
     scheduler_config.block_size = get_default_block_size(device);
-    // mode - vLLM or dynamic_split_fuse
-    scheduler_config.dynamic_split_fuse = dynamic_split_fuse;
-    // vLLM specific params
-    scheduler_config.max_num_seqs = 2;
-    scheduler_config.enable_prefix_caching = use_prefix;
 
     // It's possible to construct a Tokenizer from a different path.
     // If the Tokenizer isn't specified, it's loaded from the same folder.
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp:37]: let's drop comments about tokenizers as they are not meaningful for the SD sample

File: samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp [hunk 7]
```diff
@@ -116,34 +30,25 @@ int main(int argc, char* argv[]) try {
...
-    ov::genai::LLMPipeline pipe(model_path, device, ov::genai::draft_model(draft_model_path, device), ov::genai::scheduler_config(scheduler_config));
-
-    if (use_prefix) {
-        std::cout << "Running inference for prefix to compute the shared prompt's KV cache..." << std::endl;
-        auto generation_results = pipe.generate(prefix_str, ov::genai::greedy());
-    }
+    // User can run main and draft model on different devices.
+    // Please, set device for main model in `LLMPipeline` constructor and in in `ov::genai::draft_model` for draft.
+    // Example to run main_model on GPU and draft_model on CPU:
+    // ov::genai::LLMPipeline pipe(main_model_path, "GPU", ov::genai::draft_model(draft_model_path, "CPU"), ov::genai::scheduler_config(scheduler_config));
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp:41]: maybe we can define 2 variables `main_model_device` and `draft_model_device` ? by default they are both CPU, but you can write that powerful GPU owners can change main device to GPU

File: samples/cpp/speculative_decoding_lm/CMakeLists.txt [hunk 8]
```diff
@@ -9,26 +9,25 @@ if(POLICY CMP0135)
...
-
-find_package(OpenVINO REQUIRED COMPONENTS Runtime)
+find_package(OpenVINOGenAI REQUIRED
+    PATHS
+        "${CMAKE_BINARY_DIR}"  # Reuse the package from the build.
+        ${OpenVINO_DIR}  # GenAI may be installed alogside OpenVINO.
+    NO_CMAKE_FIND_ROOT_PATH
+)
 
 # end of dependencies
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/CMakeLists.txt:19]: you need to drop all the code starting from `# start of dependencies` to `# end of dependencies`

**Review by ilya-lavrenov — APPROVED** [review]

File: samples/cpp/speculative_decoding_lm/CMakeLists.txt [hunk 9]
```diff
@@ -1,34 +1,23 @@
...
+find_package(OpenVINOGenAI REQUIRED
+    PATHS
+        "${CMAKE_BINARY_DIR}"  # Reuse the package from the build.
+        ${OpenVINO_DIR}  # GenAI may be installed alogside OpenVINO.
+    NO_CMAKE_FIND_ROOT_PATH
+)
 
 set(TARGET_NAME speculative_decoding_lm)
 add_executable(${TARGET_NAME} ${TARGET_NAME}.cpp)
 target_link_libraries(${TARGET_NAME} PRIVATE openvino::genai cxxopts::cxxopts)
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/CMakeLists.txt:13]:
```suggestion
target_link_libraries(${TARGET_NAME} PRIVATE openvino::genai)
```

**Pull request merged**